File: parse/src/main/java/com/parse/AbstractQueryController.java
/*
 * Copyright (c) 2015-present, Parse, LLC.
 * All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree. An additional grant
 * of patent rights can be found in the PATENTS file in the same directory.
 */
package com.parse;

import com.parse.boltsinternal.Task;

/**
 * {@code AbstractParseQueryController} is an abstract implementation of
 * {@link ParseQueryController}, which implements {@link ParseQueryController#getFirstAsync}.
 */
abstract class AbstractQueryController implements ParseQueryController {

    @Override
    public <T extends ParseObject> Task<T> getFirstAsync(ParseQuery.State<T> state, ParseUser user,
                                                         Task<Void> cancellationToken) {
        return findAsync(state, user, cancellationToken).continueWith(task -> {
            if (task.isFaulted()) {
                throw task.getError();
            }
            if (task.getResult() != null && task.getResult().size() > 0) {
                return task.getResult().get(0);
            }
            throw new ParseException(ParseException.OBJECT_NOT_FOUND, "no results found for query");
        });
    }
}

File: parse/src/main/java/com/parse/AbstractQueryController.kt
/*
 * Copyright (c) 2015-present, Parse, LLC.
 * All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree. An additional grant
 * of patent rights can be found in the PATENTS file in the same directory.
 */
package com.parse

import com.parse.boltsinternal.Task

/**
 * `AbstractParseQueryController` is an abstract implementation of
 * [ParseQueryController], which implements [ParseQueryController.getFirstAsync].
 */
internal abstract class AbstractQueryController : ParseQueryController {
    override fun <T : ParseObject> getFirstAsync(
        state: ParseQuery.State<T>, user: ParseUser?,
        cancellationToken: Task<Void>?
    ): Task<T> {
        return findAsync(state, user, cancellationToken).continueWith { task: Task<List<T>?> ->
            if (task.isFaulted) {
                throw task.error
            }

            task.result?.let {
                if (it.isNotEmpty()) {
                    return@continueWith it[0]
                }
            }

            throw ParseException(ParseException.OBJECT_NOT_FOUND, "no results found for query")
        }
    }
}